### ISSUE TITLE
error with feature maps saving

### ISSUE THREAD

**Issue by viridyu** [Kamnitsask/deepmedic#35]
Hi guys

I tried to save the feature maps from the last layer of normal pathway and also subsampled pathway. Therefore I change the testConfig.cfg like this:
minMaxIndicesOfFmsToSaveFromEachLayerOfNormalPathway = [[],[],[],[],[],[],[],[0,50]]
minMaxIndicesOfFmsToSaveFromEachLayerOfSubsampledPathway = [[],[],[],[],[],[],[],[0,50]]
minMaxIndicesOfFmsToSaveFromEachLayerOfFullyConnectedPathway = [[],[],[]]

However, it went wrong like this:

Saving the new Fm-activation image for the subject #0, pathway:0, layer: 7 FM: 0
Traceback (most recent call last):
  File "./deepMedicRun", line 110, in <module>
    deepMedicTest.deepMedicTestMain(absPathToTestConf, absPathToCnnModelGivenInCmdLine)
  File "/home/by354/mywork/testcode/Segmentation_myimplement/deepmedic/deepmedic/frontEndModules/deepMedicTest.py", line 313, in deepMedicTestMain
    performInferenceForTestingOnWholeVolumes(*testSessionParameters.getTupleForCnnTesting())
  File "/home/by354/mywork/testcode/Segmentation_myimplement/deepmedic/deepmedic/trainValidateTestVisualiseParallel.py", line 1822, in performInferenceForTestingOnWholeVolumes
    myLogger
  File "/home/by354/mywork/testcode/Segmentation_myimplement/deepmedic/deepmedic/genericHelpers.py", line 273, in saveFmActivationImageToANewNiiWithHeaderFromOther
    if listOfNamesToGiveToPredictions[image_i].endswith(".nii.gz") :
IndexError: list index out of range

Is anyone meet with the similar problem? How you fixed it out? Thank you.

**Comment by Kamnitsask** [comment]
Hello,

haven't run this part of the code for a while. I ll check it in a few hours to see if I ve introduced a bug in latest versions without noticing.

In the meantime, double check something for me:
- "listOfNamesToGiveToPredictions" in the last line of this error message is a list, which contains the "names" to save the feature maps with. It's also the same list that is used for saving the segmentations with during training. That list is read from the testConfig file, given by you, the user, at the variable "namesForPredictionsPerCase = "./testNamesOfPredictions.cfg"". Please make sure this is given, and has as many entries as the number of subjects you are testing on. ie as many entries as the files you put the paths to your input-channel-files in. Does the same testing routine save the segmentation maps correctly?

I ll update this soon. Let me know that you ve set the above fine. Also, please send me the log file of your testing session to konstantinos.kamnitsas12@ic.ac.uk if you dont mind, it could help me a lot to find the problem.

K

**Comment by Kamnitsask** [comment]
Hello,

I found a bug that is probably connected to this. Have you by any chance set in the "testConfig.cfg" that you used, the following parameter to contain just Falses?

"saveProbMapsForEachClass = [False, False, False, False, False]

If you have, we found your problem. This is a totally correct option (all False), hadn't I done a very random mistake in: deepmedic.frontEndModules.frontEndHelpers.testParametersClass.py -> _makeFilepathsForPredictionsAndFeatures(...)

The problem can be fixed if you replace the contents of the function with:

```python
for case_i in xrange(self.numberOfCases) :
	if self.namesToSavePredictionsAndFeatures <> None :
		filepathForCasePrediction = absPathToFolderForPredictionsFromSession + "/" + self.namesToSavePredictionsAndFeatures[case_i]
		self.filepathsToSavePredictionsForEachPatient.append( filepathForCasePrediction )
		filepathForCaseFeatures = absPathToFolderForFeaturesFromSession + "/" + self.namesToSavePredictionsAndFeatures[case_i]
		self.filepathsToSaveFeaturesForEachPatient.append( filepathForCaseFeatures )
```

If you don't want to touch the code, in testConfig.cfg, just set any element of the"saveProbMapsForEachClass = [ .... ]" to True, and then the FMs should be saved. (this variable should be related to the FMs, if it wasnt for a bug in the logic inside the above for loop.) 

I will fix the code on the master first thing tomorrow.

Let me know if you try the above and it doesnt fix your problem. Then it will mean it's something else for you and we need to keep looking.

Sorry for the inconvenience this may have caused you.
K
